MAINT: Split out preprocessed logic into separate class

tnatt opened the pull request:
This PR adresses #593 and splits out all the logic of dealing with re-export of preprocessed data in a FMU run from `ExportData` to a new class `ExportPreprocessedData.` The re-export is mainly a copy process of the preprocessed data down to case level, but with an update of the fields `fmu`, `file` and `tracklog` in the metadata.

```python
surfacefilepath = "share/preprocessed/maps/my_surface.gri"
ExportPreprocessedData(config, casepath).export(surfacefilepath)
```


Note the initial export of these preprocessed data should still be handled by the `ExportData` class since it shares the same logic as exporting regular data, except that setting `fmu_context="preprocessed"` will ensure the data is exported and stored inside a `share/preprocessed` folder :
```python
exp = ExportData(fmu_context="preprocessed")
```

**Upsides for having a separate class for the re-export is:**

- all logic contained in one place and not spread inside `ExportData`
- users only have arguments available that will influence the process (most of the `ExportData` arguments would not impact the re-export of data)
- no need to have logic for populating the `ObjDataProvider` from existing metadata

**Downsides**

- No longer possible to update `parent/name/tagname/subfolder` when re-exporting preprocessed data 
  -  Not expected to be an issue since the common way to re-export is to loop over all files in the share/preprocessed folder without giving in these arguments.
- New class for the users to know about, but it can be informed to the users if they provide an file object to `ExportData`

perolavsvendsen commented:
I think/guess/assume that the re-export is mainly not directly user-facing? E.g. built into centrally managed components. Or do we expect that end-end users have to relate to this much?

It sounds like a valuable upside to simplify/streamline the `ExportData` class.

Not sure I see the big benefit of not having `_preprocessed` in the (temporary) metadata. Some sort of flag that tells us that this is temporary output feels useful. But I guess the point here is that when invoking the new class, this is assumed.

mferrera commented on src/fmu/dataio/preprocessed.py:
```diff
@@ -0,0 +1,226 @@
+    def _get_relative_export_path(self, existing_path: Path) -> Path:
+        """Get an updated relative_path from an existing path."""
+
+        existing_share_folder = ShareFolder.PREPROCESSED.value + "/"
+        existing_subfolders_and_filename = str(existing_path).rsplit(
+            existing_share_folder, maxsplit=1
+        )[-1]
```
I think this is fine 😄

mferrera commented on src/fmu/dataio/preprocessed.py:
```diff
@@ -0,0 +1,226 @@
+            existing_share_folder, maxsplit=1
+        )[-1]
+
+        if not self.is_observation:
+            return Path(ShareFolder.RESULTS.value) / existing_subfolders_and_filename
+        return Path(ShareFolder.OBSERVATIONS.value) / existing_subfolders_and_filename
```
Consider inverting the logic to handle the single conditional case.

```suggestion
        if self.is_observation:
            return Path(ShareFolder.OBSERVATIONS.value) / existing_subfolders_and_filename
        return Path(ShareFolder.RESULTS.value) / existing_subfolders_and_filename
```

mferrera commented on src/fmu/dataio/preprocessed.py:
```diff
@@ -0,0 +1,226 @@
+        meta_existing["file"] = self._get_meta_file(objfile, checksum_md5_file)
+
+        # update the tracklog block
+        tracklog_entry = generate_meta_tracklog(event="merged")
+        meta_existing["tracklog"].extend(tracklog_entry)
```
I think we should consider a class or set of functions to do these operations, but not in this PR #662

mferrera commented on src/fmu/dataio/preprocessed.py:
```diff
@@ -0,0 +1,226 @@
+    # Public methods:
+    # ==================================================================================
+
+    def generate_metadata(self, obj: Union[str, Path]) -> dict:
```
```suggestion
    def generate_metadata(self, obj: str | Path) -> dict:
```

mferrera commented on src/fmu/dataio/preprocessed.py:
```diff
@@ -0,0 +1,226 @@
+            f"Can't detect existing metadata file with name: {objmetafile}"
+        )
+
+    def export(self, obj: Union[str, Path]) -> str:
```
```suggestion
    def export(self, obj: str | Path) -> str:
```

mferrera reviewed (commented):
On a first pass this looks very nice, nice work. I have no significant comments but will need to take a second look.

jcrivenaes commented on src/fmu/dataio/dataio.py:
```diff
@@ -787,16 +763,24 @@ def generate_metadata(
+        if isinstance(obj, (str, Path)):
+            assert self.casepath is not None
+            _future_warning_preprocessed()
+            return deepcopy(
```
Here you return a  `deepcopy()` but later (in `export()` not. Any reasons for that?

tnatt replied:
Mainly because the export returns a `str` with the path to the exported file, while the `generate_metadata` returns a dictionary with the metadata.

jcrivenaes commented:
> Not sure I see the big benefit of not having `_preprocessed` in the (temporary) metadata. Some sort of flag that tells us that this is temporary output feels useful. But I guess the point here is that when invoking the new class, this is assumed.

I agree to keep this, so we have a unique identifier that that the intention is preprocessed context.

tnatt commented on src/fmu/dataio/preprocessed.py:
```diff
@@ -0,0 +1,226 @@
+    """
+
+    config: dict
+    casepath: Union[str, Path]
```
completely true, I keep getting confused around these 🙂

tnatt commented on src/fmu/dataio/preprocessed.py:
```diff
@@ -0,0 +1,226 @@
+
+    def __post_init__(self) -> None:
+        if get_fmu_context_from_environment() != FmuContext.CASE:
+            raise ValueError(
```
makes more sense yes 👍

tnatt commented on src/fmu/dataio/providers/_fmu.py:
```diff
@@ -104,7 +104,7 @@ class FmuProvider:
     model: dict | None = None
     fmu_context: FmuContext = FmuContext.REALIZATION
-    include_ertjobs: bool = True
+    include_ertjobs: bool = False
```
hehe nice catch.. yes was going to make a comment regarding this 🙂. I saw that in `ExportData` this is defaulted to `False` by an class variable. So I figured changing the default here were appropriate, and then I did not expose this as an argument when initializing the `FmuProvider` in the `ExportPreprocessedData`.
If I am not mistaken I think @perolavsvendsen mentioned that storing of the jobs.json per data might not be the desired way in the future? Am I correct @perolavsvendsen?

perolavsvendsen replied:
I think we stopped including the `jobs.json` quite quickly, yes. It becomes very, very large in some cases. Spamming the metadata. Rather include it in the future if there is a (more) clear need for it.

perolavsvendsen replied:
In case that wasn't very clear: As far as I am concerned, we can stop supporting this completely. Not sure if we have other dependencies on _reading_ it, but I am quite certain we can completely stop support dumping it to metadata.

tnatt replied:
> In case that wasn't very clear: As far as I am concerned, we can stop supporting this completely. Not sure if we have other dependencies on _reading_ it, but I am quite certain we can completely stop support dumping it to metadata.

I like the sound of that 👍

tnatt replied:
created issue #664

tnatt commented:
> I think/guess/assume that the re-export is mainly not directly user-facing? E.g. built into centrally managed components. Or do we expect that end-end users have to relate to this much?

Yes this is correct, a regular user will not be exposed to this. In Drogon it is used inside the ERT workflow `WF_COPY_PREPROCESSED_DATAIO`, which probably many users have copied and not modified at all. Other than that export of preprocessed data is being used e.g. inside the ongoing `SIM2SEIS` and `genertobs` work.

This local ERT workflow in Drogon `WF_COPY_PREPROCESSED_DATAIO` has a note stating that this is temporarily until available through `fmu-dataio` as a pre-installed ERT workflow.  It is this temporary solution that made it necessary to initialize the `ExportPreprocessedData` inside the `ExportData` class, as this local script probably will live for quite some time before being replaced by a newer script or a pre-installed ERT workflow (knowledge from experience 😄)

perolavsvendsen commented on src/fmu/dataio/dataio.py:
```diff
@@ -48,12 +49,22 @@
+    warnings.warn(
+        "Using the ExportData class for re-exporting preprocessed data is no "
+        "longer supported. Use the dedicated ExportPreprocessedData class "
+        "instead. In a deprecation period the ExportPreprocessedData is used "
```
Not the most critical comment... But is it necessary to state here that `ExportPreprocessedData` is used under the hood?

tnatt replied:
probably not, but I was mainly saying it in the rare case they experienced a change in behaviour.. The `ExportPreprocessedData` class does not support updating filenames hence if users specify e.g. the argument `name` in `ExportData` it will not be populated into `ExportPreprocessedData`. 

I don't reckon anyone has done that though.. probably the users only have the ERT workflow `WF_COPY_PREPROCESSED_DATAIO`, and no argument is given in there to update the file name on the re-export.

perolavsvendsen commented on src/fmu/dataio/dataio.py:
```diff
@@ -48,12 +49,22 @@
+        "longer supported. Use the dedicated ExportPreprocessedData class "
+        "instead. In a deprecation period the ExportPreprocessedData is used "
+        "under the hood when a filepath is input to ExportData. "
+        "This will change in the future!",
```
"Please update your script, as this will be discontinued in the future."? (call to action)

tnatt commented on src/fmu/dataio/_metadata.py:
```diff
@@ -37,12 +32,14 @@
 
 
-def generate_meta_tracklog() -> list[meta.TracklogEvent]:
+def generate_meta_tracklog(
+    event: Literal["created", "merged"] = "created",
+) -> list[meta.TracklogEvent]:
```
A bit unsure if it is two or three types. In the `meta.TracklogEvent` this field is typed as a `str` and the examples states three types. I can't see any reference to this `updated` event inside `dataio` so I'm quite certain dataio don't create metadata where `tracklog.event = "updated"`.

https://github.com/equinor/fmu-dataio/blob/0c2d04016875c2194b0f1777c37bb743e5880e1b/src/fmu/dataio/datastructure/meta/meta.py#L327-L329

Probably we should constrain the type for the event field inside the `meta.TracklogEvent` to an `Enum` or at least a `Literal` instead of `str`

tnatt replied:
Created #669

tnatt commented:
I have updated the PR with tests and added back the `_preprocessed` flag but now in the form of a bool 🙂 While working on the tests I've added a bit more checks and messages to the code.

Great if some of you find the time to review again @mferrera @jcrivenaes @perolavsvendsen

perolavsvendsen reviewed (approved):
🚀
